crossbeam-channel/src/flavors/array.rs

```diff
@@ -527,6 +534,32 @@ impl<T> Channel<T> {
     }
 }
 
+impl<T: ToOwned<Owned = T>> Channel<T> {
+    /// Writes a message into the channel without cloning if writing fails.
+    pub(crate) unsafe fn write_cow<'a>(&self, token: &mut Token, msg: Cow<'a, T>) -> Result<(), Cow<'a, T>> {
+        // If there is no slot, the channel is disconnected.
+        if token.array.slot.is_null() {
+            return Err(msg);
+        }
+
+        // Take ownership of/clone the underlying message.
+        let msg = msg.into_owned();
```

Comment:
Panicking here would be bad because it would stall the channel because `write_unchecked` is never called.

Reply:
I'm not too familiar on catching/handling panicking in Rust yet. Should we use these?

https://doc.rust-lang.org/std/panic/trait.UnwindSafe.html

https://doc.rust-lang.org/std/panic/fn.catch_unwind.html

Reply:
A better pattern for this case is RAII scope guard, such as [scopeguard](https://docs.rs/scopeguard/1/scopeguard) (because it has no additional cost).